### ISSUE TITLE
Changing TF version changes user-data hash

### ISSUE THREAD

**Issue by dbogen** [hashicorp/terraform#13723]
We were running with version 0.8.6.  I decided to try and upgrade to the most recent version to take advantage of some of the whiz-bang new features.

However, upon doing so, TF informed me that it would replace every instance in our AWS fleet because it calculated a change in user-data.  This, despite doing absolutely nothing to the TF files.  The *only* thing that changed was the TF version.

Further investigation revealed that I could safely upgrade to 0.8.7, but 0.8.8 or the current version, 0.9.3, would propose a fleet replacement.

### Terraform Version
0.8.8 and above

### Affected Resource(s)
- aws_instance
- template_file

### Terraform Configuration Files
Sample configuration file below that reliably demonstrates the behavior I see.

```
resource "aws_instance" "test_server_1" {
  ami                         = "ami-fce3c696"
  availability_zone           = "us-east-1a"
  ebs_optimized               = false
  instance_type               = "t2.micro"
  subnet_id                   = "${aws_subnet.public_subnet.id}"
  associate_public_ip_address = false
  user_data                   = "${data.template_cloudinit_config.test_user_data_1.rendered}"
  vpc_security_group_ids = [
    "${aws_security_group.test_server.id}",
  ]
}

data "template_cloudinit_config" "test_user_data_1" {
  part {
    content_type = "text/x-shellscript"
    content      = "${data.template_file.simple_echo.rendered}"
  }
}

data "template_file" "simple_echo" {
  template = "#!/bin/bash\necho \"$${foo}\""

  vars {
    foo = "bar"
  }
}

provider "aws" {
  region = "us-east-1"
}

resource "aws_vpc" "default" {
  cidr_block           = "10.0.0.0/16"
  enable_dns_hostnames = true
  enable_dns_support   = true

  tags {
    Name = "test-environment"
  }
}

resource "aws_subnet" "public_subnet" {
  vpc_id = "${aws_vpc.default.id}"

  cidr_block        = "10.0.1.0/24"
  availability_zone = "us-east-1a"

  tags {
    Name = "public-subnet"
  }
}

resource "aws_security_group" "test_server" {
  name        = "test_server"
  vpc_id      = "${aws_vpc.default.id}"
  description = "Applied to test server instances."

  ingress {
    from_port = 22
    to_port   = 22
    protocol  = "tcp"
    self      = true
  }

  egress {
    from_port   = 0
    to_port     = 0
    protocol    = "-1"
    cidr_blocks = ["0.0.0.0/0"]
  }
}
```

### Debug Output
https://gist.github.com/dbogen/4b0864b93a435d8aec9f68554754ecac
Note that account numbers and usernames have been changed.

### Expected Behavior
Nothing.  Doing a `terraform plan` with 0.8.8 or above should produce no changes if the TF configuration files don't change.

### Actual Behavior
Changing from 0.8.7 to 0.8.8 or above (including 0.9.3) causes terraform to calculate completely different hash values for my instance user-data.  That causes TF to plan on replacing our entire AWS instance fleet.

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply` using version 0.8.7.
2. Upgrade to version 0.8.8 or above
3. `terraform plan` with the new version.  Notice that the hash on the user-data has changed and a replacement of the instance is proposed.

### Important Factoids
N/A

### References
N/A

**Comment by johnnycarlin** [comment]
Using `ignore_changes` as a temporary workaround.

```
lifecycle {
  ignore_changes = ["user_data"]
}
```

**Comment by patdowney** [comment]
diffing the user data before and after the upgrade from 0.8.7 to 0.8.8 the only think that I could see different is 4 bytes at the begging of the gzipped file that are used to store a modification time - which led me to this bug (https://github.com/golang/go/issues/17663) in the `compress/gzip` library that was fixed in go1.8 (https://github.com/golang/go/commit/b4e714e59efd21f9140a3e8e2b0182c74bfde437).  

We're experiencing this bug as terraform 0.8.8 has been compiled with 1.8 and includes this fix, whereas versions between 0.7.11 and .0.8.7 were compiled with version 1.7.3 which would still have had the bug.

I know this doesn't fix the issue at hand - we ended up using the `ignore_changes` workaround mentioned above.

**Comment by apparentlymart** [comment]
Hmm... thanks for that pointer @patdowney!

Indeed it sounds like `template_cloudinit_config` has changed its behavior due to the change to the underlying gzip library. Fixing this is likely to require some trickery, such as having the `user_data` attribute sniff to see if its content looks like it has a `gzip` header and ignore the diff if only the timestamp field seems to have changed.

In the mean time, `ignore_changes` does indeed seem like the best workaround.
